spark-version/1.5-plus/src/main/scala/ZippedPartitionsWithLocalityRDD.scala

```diff
@@ -62,15 +62,17 @@ class ZippedPartitionsWithLocalityRDD[A: ClassTag, B: ClassTag, V: ClassTag](
       while (p < candidateLocs.length) {
         locs = candidateLocs(p)._2.intersect(curPrefs)
         if (!locs.isEmpty) {
-          matchPartition = candidateLocs.apply(p)
+          matchPartition = candidateLocs.remove(p)
           p = Integer.MAX_VALUE - 1
         }
         p += 1
       }
-      require(matchPartition != null, s"can't find locality partition for partition $i " +
-        s"Partition locations are (${curPrefs}) Candidate partition locations are\n" +
-        s"${candidateLocs.mkString("\n")}. Are you using more executors than the node number? " +
-        s"If yes, try to change your executor number.")
+      if (matchPartition == null) {
+        println(s"can't find locality partition for partition $i " +
+          s"Partition locations are (${curPrefs}) Candidate partition locations are\n" +
+          s"${candidateLocs.mkString("\n")}.")
+        matchPartition = candidateLocs.remove(util.Random.nextInt(candidateLocs.length))
```

Comment:
While this would work, it may randomly pick a partition (from candidateLocs) that has common location with a (yet to be processed) partition from rdd(0); a better solution is:
1) First zip partitions with common locations as much as we can;
2) After we have zipped all the partitions with common locations, we can then process the remaining partitions (which have no common locations with others); we can simply zip them together in order (just as ZippedPartitionsBaseRDD does).